shared/src/balancer_event_handler.rs

```diff
@@ -0,0 +1,198 @@
+use crate::{
+    current_block::BlockRetrieving,
+    event_handling::{BlockNumber, EventHandler, EventIndex, EventStoring},
+    impl_event_retrieving,
+    maintenance::Maintaining,
+};
+use anyhow::{anyhow, Context, Result};
+use contracts::{
+    balancer_v2_vault::{
+        self, event_data::PoolRegistered as ContractPoolRegistered, Event as ContractEvent,
+    },
+    BalancerV2Vault,
+};
+use ethcontract::common::DeploymentInformation;
+use ethcontract::{dyns::DynWeb3, Event as EthContractEvent, EventMetadata, H160, H256};
+use std::fmt::{Debug, Formatter};
+use std::ops::RangeInclusive;
+use tokio::sync::Mutex;
+
+#[derive(Debug)]
+pub enum BalancerEvent {
+    PoolRegistered(PoolRegistered),
+}
+
+#[derive(Debug)]
+pub struct PoolRegistered {
+    pub pool_id: H256,
+    pub pool_address: H160,
+    pub specialization: PoolSpecialization,
+}
+
+/// There are three specialization settings for Pools, which allow for cheaper swaps at the cost of reduced
+/// functionality:
+///
+///  - General: no specialization, suited for all Pools. IGeneralPool is used for swap request callbacks, passing the
+/// balance of all tokens in the Pool. These Pools have the largest swap costs (because of the extra storage reads),
+/// which increase with the number of registered tokens.
+///
+///  - Minimal Swap Info: IMinimalSwapInfoPool is used instead of IGeneralPool, which saves gas by only passing the
+/// balance of the two tokens involved in the swap. This is suitable for some pricing algorithms, like the weighted
+/// constant product one popularized by Balancer V1. Swap costs are smaller compared to general Pools, and are
+/// independent of the number of registered tokens.
+///
+///  - Two Token: only allows two tokens to be registered. This achieves the lowest possible swap gas cost. Like
+/// minimal swap info Pools, these are called via IMinimalSwapInfoPool.
+#[repr(u8)]
+pub enum PoolSpecialization {
+    General = 0,
+    MinimalSwapInfo = 1,
+    TwoToken = 2,
+}
+
+impl PoolSpecialization {
+    fn new(specialization: u8) -> Result<Self> {
+        match specialization {
+            0 => Ok(Self::General),
+            1 => Ok(Self::MinimalSwapInfo),
+            2 => Ok(Self::TwoToken),
+            t => Err(anyhow!("Invalid PoolSpecialization value {}", t)),
+        }
+    }
+}
+
+impl std::fmt::Debug for PoolSpecialization {
```

Comment:
Why not just `derive(Debug)`?

AFAIK it just works:
```rust
#[derive(Debug)]
#[repr(u8)]
enum Foo {
    Bar = 0,
    Baz = 1,
}

fn main() {
    println!("{:?} and {:?}", Foo::Bar, Foo::Baz); // prints "Bar and Baz"
}
```

Reply:
I totally tried that, but maybe it was the ordering that didn't work.

I had
```
#[repr(u8)]
#[derive(Debug)]
```

Will try again

Reply:
Just checked an order shouldn't matter: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b0ffd9b7738412ad007638858ee4c1cc